### PR TITLE
Packaging

### LINKED ISSUE #8: Integrate code into package/s

### PR DESCRIPTION
**Pull request by vnoelifant**
Followed https://packaging.python.org/en/latest/tutorials/packaging-projects/ to move source code to a package. Also watched some of https://pybites.mykajabi.com/products/pybites-coaching-calls-2/categories/3469727/posts/2166525740 to learn how to run code with package and typer.  I haven't gotten to the part of creating the toml file though, that will be next steps I believe.  I was able to run everything with `python -m src.cozmo_companion speech` where "speech" is the name of the wav file I have selected.

### CODE REVIEW

**Comment by vnoelifant** [comment]
@bbelderbos Wow, Typer is cool, just started using it and reading about it. No more exception block code and sys.arg for arguments! 
Screenshot (510)

**Comment by vnoelifant** [comment]
@bbelderbos What is the preferred way to create an output path for the speech.wav to an output folder? I have the following in https://github.com/vnoelifant/cozmo-companion/blob/packaging/src/cozmo_companion/__main__.py: 

```
   curr_dir = os.getcwd()
   speech_out = os.path.join(curr_dir, "wav_output", filename + ".wav"
```

**Comment by vnoelifant** [comment]
@bbelderbos When are the dots before the module name not needed if you import modules from current directory? Do you only need the dots if you have code in a package? 
```
from .dialogue_cozmo import Dialogue
from .recorder_cozmo import Recorder
```

**Comment by vnoelifant** [comment]
@bbelderbos Is it really necessary to have the `if __name__ == "__main__":` in https://github.com/vnoelifant/cozmo-companion/blob/packaging/src/cozmo_companion/__main__.py?

**Comment by bbelderbos** [comment]
> @bbelderbos What is the preferred way to create an output path for the speech.wav to an output folder? I have the following in https://github.com/vnoelifant/cozmo-companion/blob/packaging/src/cozmo_companion/__main__.py:
> 
> ```
>    curr_dir = os.getcwd()
>    speech_out = os.path.join(curr_dir, "wav_output", filename + ".wav"
> ```

That works but I would recommend using pathlib.Path

**Comment by bbelderbos** [comment]
> 

Exactly, if you just have mod1.py and mod2.py without package, you can do import mod1 in mod2.py, but when it's a package you do relative or absolute imports, the .module is a relative import, you can also do "from package.module" as an absolute import. I will review this shortly ...

**Comment by bbelderbos** [comment]
> @bbelderbos Is it really necessary to have the `if __name__ == "__main__":` in https://github.com/vnoelifant/cozmo-companion/blob/packaging/src/cozmo_companion/__main__.py?

Good question, Will actually adviced me against it as per docs, so lately I do this in __main__:

```
from .cli import app

app()
```

(example from pybites-search)

**Comment by bbelderbos** [comment]
> @bbelderbos When are the dots before the module name not needed if you import modules from current directory? Do you only need the dots if you have code in a package?
> 
> ```
> from .dialogue_cozmo import Dialogue
> from .recorder_cozmo import Recorder
> ```

> 

Exactly, if you just have mod1.py and mod2.py without package, you can do import mod1 in mod2.py, but when it's a package you do relative or absolute imports, the .module is a relative import, you can also do "from package.module" as an absolute import. I will review this shortly ...

File: src/cozmo_companion/__main__.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import json
+import typer
+import sys
+import os
+import traceback
+from .dialogue_cozmo import Dialogue
```
Comment (bbelderbos) [inline src/cozmo_companion/__main__.py]: yes as cozmo_companion is a package, this is the way to go

File: src/cozmo_companion/__main__.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import json
+import typer
```
Comment (bbelderbos) [inline src/cozmo_companion/__main__.py]: check PEP8 for import organization ...

File: src/cozmo_companion/__main__.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            print("Transcribing audio....\n")
+
+            try:
+                speech_text = my_convo.transcribe_audio()
+                my_convo.get_cozmo_response(speech_text)
+                if "exit" in speech_text:
+                    print("Exiting program...")
+                    break
+
+            except:
```
Comment (bbelderbos) [inline src/cozmo_companion/__main__.py]: What exception?

File: src/cozmo_companion/__main__.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            print("starting recording process")
+            recorder = Recorder(speech_out)
+            print("Please say something nice into the microphone\n")
+            recorder.save_to_file()
+            print("Transcribing audio....\n")
+
+            try:
+                speech_text = my_convo.transcribe_audio()
+                my_convo.get_cozmo_response(speech_text)
+                if "exit" in speech_text:
```
Comment (bbelderbos) [inline src/cozmo_companion/__main__.py]: What if exit is not in text, would the loop ever break? Or are we ok with the keyboard interrupt signal you do catch below?

File: src/cozmo_companion/cozmo_echo.py [hunk 5]
```diff
@@ -23,7 +23,7 @@ def main():
             try:
                 speech_text = my_convo.transcribe_audio()
                 my_convo.get_cozmo_response(speech_text)
-                if "goodbye" in speech_text or "bye" in speech_text:
+                if "exit" in speech_text:
```
Comment (bbelderbos) [inline src/cozmo_companion/cozmo_echo.py]: "exit" hardcoded twice, make a constant and import it?

File: speech.wav
Comment (bbelderbos) [inline speech.wav:1]: seems you need to gitignore this file?

File: src/cozmo_companion/__main__.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
+import json
+import typer
+import sys
+import os
+import traceback
+from .dialogue_cozmo import Dialogue
+from .recorder_cozmo import Recorder
+
+
+def main(filename: str):
```
Comment (bbelderbos) [inline src/cozmo_companion/__main__.py]: so can only be called with command line arg right?

**Review by bbelderbos — APPROVED** [review]
Looks good, couple of comments / questions. As you're working on packaging, you still need to include a pyproject.toml file right?

File: requirements.txt [hunk 7]
```diff
@@ -1,17 +1,35 @@
+black==23.3.0
 certifi==2022.12.7
 charset-normalizer==3.1.0
+click==8.1.3
```
Comment (bbelderbos) [inline requirements.txt:4]: you can also manage those in your pyproject.toml you'll add for packaging, example: https://github.com/PyBites-Open-Source/search/blob/main/pyproject.toml

**Comment by bbelderbos** [comment]
> > @bbelderbos What is the preferred way to create an output path for the speech.wav to an output folder? I have the following in https://github.com/vnoelifant/cozmo-companion/blob/packaging/src/cozmo_companion/__main__.py:
> > ```
> >    curr_dir = os.getcwd()
> >    speech_out = os.path.join(curr_dir, "wav_output", filename + ".wav"
> > ```
> 
> That works but I would recommend using pathlib.Path

Example:

```
>>> Path.cwd() / "wav_output" / f"{filename}.wav"
PosixPath('/Users/bbelderbos/code/PDM/projects/ronnie/cozmo-companion/wav_output/output_file.wav')
```

(sub directory does need to exist for that to work, if not you can create it by the script)

**Comment by bbelderbos** [comment]
> > @bbelderbos Is it really necessary to have the `if __name__ == "__main__":` in https://github.com/vnoelifant/cozmo-companion/blob/packaging/src/cozmo_companion/__main__.py?
> 
> Good question, Will actually adviced me against it as per docs, so lately I do this in **main**:
> 
> ```
> from .cli import app
> 
> app()
> ```
> 
> (example from pybites-search)

Reason of not using if name == main in __main__.py:

- [__main__.py idiomatic usage](https://docs.python.org/3/library/__main__.html#id1) - reason we ditched the if __name__ == "__main__": from __main__.py

Also discussed in this clinic: https://pybites.mykajabi.com/products/pybites-coaching-calls-2/categories/3469727/posts/2166525740

Refactoring suggestion: move the code you currently have in __main__ into cli.py and use the from .cli import app code above, see pybites-search for example: https://github.com/PyBites-Open-Source/search/blob/main/src/pybites_search/cli.py

File: src/cozmo_companion/__main__.py (hunk 2 above)
Comment (bbelderbos) [reply src/cozmo_companion/__main__.py]: running isort it changes it to the recommended way:

```
import json
import os
import sys
import traceback

import typer

from .dialogue_cozmo import Dialogue
from .recorder_cozmo import Recorder
```

See also https://pep8.org/#imports

File: src/cozmo_companion/__main__.py (hunk 4 above)
Comment (bbelderbos) [reply src/cozmo_companion/__main__.py:28]: could timeout be a thing here?

File: src/cozmo_companion/__main__.py (hunk 3 above)
Comment (bbelderbos) [reply src/cozmo_companion/__main__.py]: so what exceptions could transcribe_audio() and get_cozmo_response() raise?

if those are your functions, you can also raise your own exceptions (https://codechalleng.es/tips/defining-custom-exceptions), for example:

```
def get_user_model():
    """
    Return the User model that is active in this project.
    """
    try:
        return django_apps.get_model(settings.AUTH_USER_MODEL, require_ready=False)
    except ValueError:
        raise ImproperlyConfigured("AUTH_USER_MODEL must be of the form 'app_label.model_name'")
    except LookupError:
        raise ImproperlyConfigured(
            "AUTH_USER_MODEL refers to model '%s' that has not been installed" % settings.AUTH_USER_MODEL
        )
```
(from django)

By using "custom exception" `ImproperlyConfigured` the caller of this function only has to deal with that one exception, e.g.

```
try:
    get_user_model()
except ImproperlyConfigured:
    pass
```

**Pull request merged**